apps/language_server/lib/mix/tasks/lsp/mappings/generate.ex

```diff
@@ -0,0 +1,223 @@
+      (defaults to `ElixirLS.LanguageServer.Experimental.Protocol.Proto`)
+    * `--only` - Only generate the LSP types in the comma separated list
+    * `--roots` - A comma separated list of types to import. The types given will be interrogated
+      and all their references will also be imported. This is useful when importing complex structures,
+      as you don't need to specify all the types you wish to import.
+  """
+
+  alias Mix.Tasks.Lsp.DataModel
+  alias Mix.Tasks.Lsp.Mappings
+  alias Mix.Tasks.Lsp.Mappings.Mapping
+
+  @generated_files_root ~w(lib generated)
+  @switches [
+    only: :string,
+    proto_module: :string,
+    roots: :string,
+    types_module: :string
+  ]
+
+  def run(args) do
+    args
+    |> parse_options()
+    |> do_run()
+  end
+
+  def do_run(options) do
+    mappings_opts = Keyword.take(options, [:types_module, :proto_module])
+
+    with {:ok, %DataModel{} = data_model} <- DataModel.new(),
+         {:ok, mappings} <- Mappings.new(mappings_opts),
+         {:ok, types_to_map} <- get_mapped_types(options, data_model),
+         {:ok, results} <- map_lsp_types(types_to_map, data_model, mappings) do
+      IO.puts("Mapping complete, writing #{length(results)} files")
+
+      for {file_name, ast} <- results do
+        write_file(file_name, ast)
+        IO.write(".")
+      end
+
+      IO.puts("\nComplete.")
+    else
+      {:error, reason} ->
+        Mix.Shell.IO.error("An error occurred during mapping #{to_string(reason)}")
+
+      error ->
+        Mix.Shell.IO.error("An error occurred during mapping #{inspect(error)}")
+    end
+  end
+
+  defp parse_options(args) do
+    {keywords, _, _} = OptionParser.parse(args, strict: @switches)
+
+    if Keyword.has_key?(keywords, :only) and Keyword.has_key?(keywords, :roots) do
+      raise "You can only specify one of --only and --roots"
+    end
+
+    keywords
+    |> Keyword.replace_lazy(:only, &split_comma_delimited/1)
+    |> Keyword.replace_lazy(:roots, &split_comma_delimited/1)
+  end
+
+  defp map_lsp_types(types_to_map, %DataModel{} = data_model, %Mappings{} = mappings) do
+    mapping_results =
+      types_to_map
+      |> Enum.map(&do_mapping(&1, mappings, data_model))
+      |> Enum.reduce_while([], fn
+        :skip, acc ->
+          IO.write([IO.ANSI.yellow(), ".", IO.ANSI.reset()])
+
+          {:cont, acc}
+
+        {:ok, file, ast}, results ->
+          IO.write([IO.ANSI.green(), ".", IO.ANSI.reset()])
+
+          {:cont, [{file, ast} | results]}
+
+        error, _ ->
+          IO.write([IO.ANSI.red(), "x", IO.ANSI.reset()])
+
+          {:halt, error}
+      end)
+
+    IO.puts("")
+
+    case mapping_results do
+      results when is_list(results) ->
+        {:ok, Enum.reverse(results)}
+
+      error ->
+        error
+    end
+  end
+
+  defp do_mapping(%struct_module{} = structure, %Mappings{} = mappings, %DataModel{} = data_model) do
+    with {:ok, %Mapping{} = mapping} <- Mappings.fetch(mappings, structure.name),
```

Comment:
`mapping` is not used. Is it an error?

Reply:
It's just there to ensure that the mapping exists. I can remove the variable.